### ISSUE TITLE
Missing TypeScript Definitions with ECMAScript Modules

### ISSUE THREAD

**Issue by qjroberts** [twurple/twurple#383]
## Bug Report

I was integrating this into an existing project, and ran into an issue importing the TypeScript definitions. It looks like there's some conflict with having `"type": "module"` in `package.json` + targeting `"es2022"` in `tsc`. Looking up some ideas, at least with my version of Node/TypeScript this can be fixed by adding `"types"` path to the `"exports"` in each `package.json`. For example, in `@twurple/auth`'s `package.json` it would look like:

```json
"exports": {
  ".": {
    "require": "./lib/index.js",
    "import": "./es/index.mjs",
    "types": "./lib/index.d.ts" // This is the only new line
  }
}
```

I can submit a PR updating these, if this seems like a reasonable solution! Also open to other ideas, and willing to work on it :)

### Code

Run `npm run repro` with the following two files:

**package.json**
```json
{
  "name": "twurple-test",
  "version": "1.0.0",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "repro": "tsc --strict --module node16 --target es2022 index.ts"
  },
  "dependencies": {
    "@twurple/auth": "^5.2.1"
  },
  "devDependencies": {
    "typescript": "^4.7.4"
  }
}
```

**index.ts**
```ts
import type { AccessToken } from "@twurple/auth";
import { RefreshingAuthProvider } from "@twurple/auth";

let tokenData: AccessToken = {
  accessToken: "",
  refreshToken: "",
  scope: [],
  expiresIn: 0,
  obtainmentTimestamp: 0,
};

export default new RefreshingAuthProvider(
  { clientId: "", clientSecret: "" },
  tokenData
);
```

> **Note:** To reproduce this, `package.json` must include `"type": "module"` (enforces `import` syntax) and `tsc` must have `strict` enabled (missing definition becomes an error), `module` must be `node16` and `target` must be `"es2022"`.

### Expected behavior
Importing classes automatically resolves type definitions.

### Actual Behavior

TypeScript strict mode error: `Could not find a declaration file for module '@twurple/auth'.`

### Environment

- Version: 5.2.1
- Node version: 18.4.0
- Operating system: Ubuntu 20.04.4 (LTS) via WSL (Windows 10)

**Comment by d-fischer** [comment]
Sure, I have not looked into recent ESM + TS development a lot, but if that makes sense to support additional compilation targets without breaking any compatibility to already working use cases, I'm all for patching it in.

**Comment by Reprevise** [comment]
Yeah I have the same issue, the only package that works is easy-bot, not too sure why though. Would be great to get that quick fix out to the rest of the packages.

**Comment by qjroberts** [comment]
> Sure, I have not looked into recent ESM + TS development a lot, but if that makes sense to support additional compilation targets without breaking any compatibility to already working use cases, I'm all for patching it in.

Sounds good - I'll throw up a PR for review in a bit. Are there any specific workflows you'd like me to verify to make sure it doesn't break existing compatibility? I don't think it should, but wouldn't hurt to check!

> Yeah I have the same issue, the only package that works is easy-bot, not too sure why though. Would be great to get that quick fix out to the rest of the packages.
> 
> EDIT: The `easy-bot` package works because the types are in the same directory as the javascript code.

Yeah, this only happens when Node tries to import the new ESM packages, since the types are all in the old "require" (`/lib`) directory. TypeScript can't find the types that correspond. There is a `types` top-level attribute, but I think that might be old/deprecated, and the `types` under `exports` is the "newer" way.

**Comment by d-fischer** [comment]
> Yeah I have the same issue, the only package that works is easy-bot, not too sure why though. Would be great to get that quick fix out to the rest of the packages.

That package can resolve types because it has no explicit exports mapping, and introducing it now would be a breaking change, so it has to wait until 6.0.

**Comment by d-fischer** [comment]
Fixed with #384 and #385, closing.
